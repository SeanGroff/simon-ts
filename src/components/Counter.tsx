import * as React from 'react';
import styled from 'styled-components';

interface Props {
  on: boolean;
  count: number;
}

const StyledCounter = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  border: 4px solid #222222;
  background-color: #32050c;
  font-size: 32px;
  border-radius: 10px;
  width: 50px;
  color: ${(props: Props) => (props.on ? '#DC0D29' : '#430710')};
`;

const Counter = (props: Props) => (
  <StyledCounter {...props}>{props.count || '--'}</StyledCounter>
);

Counter.defaultProps = {
  count: 0,
};

export default Counter;
