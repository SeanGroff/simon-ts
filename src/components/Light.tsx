import * as React from 'react';
import styled, { keyframes, StyledProps } from 'styled-components';

interface Props {
  id: number;
  animate: boolean;
  color: string;
  lightUpColor: string;
  clickable: boolean;
  topLeft?: boolean;
  topRight?: boolean;
  bottomLeft?: boolean;
  bottomRight?: boolean;
  onClick(): void;
}

interface LightUpProps {
  color: string;
  lightUpColor: string;
}

const Light: React.StatelessComponent<Props> = props => {
  console.log(props);
  return <div />;
};

const LightUp = (props: LightUpProps) => keyframes`
  0% {
    background-color: ${props.color};;
  }
  30%, 60% {
    background-color: ${props.lightUpColor};
  }
  100% {
    background-color: ${props.color};
  }
`;

const StyledLight = styled(Light)`
  display: inline-block;
  width: 200px;
  height: 200px;
  background-color: ${(props: StyledProps<Props>) => props.color};
  border: 12px solid #333333;
  border-top-left-radius: ${(props: StyledProps<Props>) =>
    props.topLeft ? '100%' : '0'};
  border-top-right-radius: ${(props: StyledProps<Props>) =>
    props.topRight ? '100%' : '0'};
  border-bottom-left-radius: ${(props: StyledProps<Props>) =>
    props.bottomLeft ? '100%' : '0'};
  border-bottom-right-radius: ${(props: StyledProps<Props>) =>
    props.bottomRight ? '100%' : '0'};
  pointer-events: ${(props: StyledProps<Props>) =>
    props.clickable ? 'auto' : 'none'};
  animation: ${(props: StyledProps<Props>) =>
    props.animate ? `${LightUp(props)} 1500ms linear` : 'none'};
  :hover {
    cursor: ${(props: StyledProps<Props>) =>
      props.clickable ? 'pointer' : 'initial'};
  }
`;

export default (props: Props) => <StyledLight {...props} />;
