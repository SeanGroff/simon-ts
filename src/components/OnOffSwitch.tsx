import * as React from 'react';
import styled from 'styled-components';

interface Props {
  power: boolean;
  powerOnOff(power: boolean): { type: string; payload: boolean };
}

const StyledSwitch = styled.div`
  width: 40px;
  height: 20px;
  background-color: #222222;

  :hover {
    cursor: pointer;
  }
`;

const Toggler = styled.div`
  background-color: #fc0102;
  width: 16px;
  height: 16px;
  border-radius: 4px;
  border: 2px solid #222222;
  position: relative;
  left: ${(props: Props) => (props.power ? '20px' : '0')};
  transition: left 0.1s ease-out;
`;

const OnOffSwitch = (props: Props) => {
  const clickHandler = () => props.powerOnOff(!props.power);
  return (
    <StyledSwitch onClick={clickHandler} {...props}>
      <Toggler {...props} />
    </StyledSwitch>
  );
};

export default OnOffSwitch;
