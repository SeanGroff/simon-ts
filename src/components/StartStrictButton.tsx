import * as React from 'react';
import styled from 'styled-components';

interface Props {
  startButton: boolean;
  disabled: boolean;
  power: boolean;
  startGame?(): { type: string };
  onClick?(): void;
}

// const Button: React.StatelessComponent<Props> = props => <button />;

const StyledButton = styled.button`
  border-radius: 100%;
  width: 20px;
  height: 20px;
  border: 4px solid #222222;
  background: ${(props: Props) => (props.startButton ? '#FC0102' : 'yellow')};
  margin: auto;
  padding: 10px;
  position: relative;
  top: -5px;
  box-shadow: 0 2px 3px #222;

  &:hover {
    cursor: pointer;
  }

  &:active {
    top: -4px;
    box-shadow: 0 1px 1px #292929;
  }
`;

const StartStrictButton = (props: Props) => {
  const clickHandler = () => {
    if (props.startButton && props.startGame !== undefined) {
      props.startGame();
    } else {
      /**
       * @todo strictGame()
       */
    }
  };
  return <StyledButton onClick={clickHandler} {...props} />;
};

export default StartStrictButton;
