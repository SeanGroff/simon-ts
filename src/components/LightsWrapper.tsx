import * as React from 'react';
import styled from 'styled-components';
import Light from './Light';
// import { getSequenceOfRandomNumbers, playSequence } from '../utils/logic';

interface Props {
  power: boolean;
  counter: number;
  lightSequence: number[];
  playerTurn: boolean;
  toggleGamePowerAction(payload: boolean): { type: string; payload: boolean };
  startGameAction(): { type: string };
  // roundSuccessThunk(): any;
  // nextTurnThunk(): any;
}

interface State {
  sequence: number[];
  clickable: boolean;
  lightOn: number;
}

const LightsWrapper = styled.div`position: relative;`;

const LightsRow = styled.div`margin-bottom: -4px;`;

export default class Lights extends React.Component<Props, State> {
  /**
   * @todo change lightOn to lightSequence
   *
   * @memberof Lights
   */
  state = {
    sequence: [],
    clickable: false,
    lightOn: 100,
  };

  constructor(props: Props) {
    super(props);
    this.clickHandler = this.clickHandler.bind(this);
    this.toggleClickable = this.toggleClickable.bind(this);
    this.playSequence = this.playSequence.bind(this);
  }

  /**
   * @todo When the component mounts GET an Array of random numbers
   *  with a size of 25 and set the lightSequence local state to that Array
   *
   * @memberof Lights
   */
  componentWillMount() {
    // this.setState(() => ({
    //   sequence: getSequenceOfRandomNumbers(),
    // }));
  }

  /**
   *
   * @todo fix this lifecycle method if statement
   *
   * @param {Props} { playerTurn, lightSequence }
   * @memberof Lights
   */
  componentWillReceiveProps({ playerTurn, lightSequence, counter }: Props) {
    /**
     * Resets the lights before the Players turn
     */
    // if (playerTurn && lightSequence && lightSequence.length) {
    //   lightSequence.forEach((light: number): void => {
    //     this.setState((prevState, props) => ({
    //       lightOn: 100,
    //     }));
    //   });
    // }
    /**
     * Sets the state to the light that should light up in the sequence
     */
    // if (
    //   !playerTurn &&
    //   counter &&
    //   this.state.sequence &&
    //   this.state.sequence.length
    // ) {
    //   this.playSequence();
    // }
  }

  clickHandler(id: number) {
    // this.lightUp(id);
    // const lightSeq = this.props.lightSequence;
    // this.setState((prevState, props) => ({
    //   lightOn: id,
    // }));
    // if (lightSeq && lightSeq.length) {
    //   const results = lightSeq.filter(light => light === id);
    //   if (results && results.length === lightSeq.length) {
    //     console.log('success!');
    //     this.props.roundSuccessThunk();
    //   } else {
    //     console.log('WRONG!');
    //     // fire INCORRECT_CHOICE action
    //   }
    // }
  }

  toggleClickable(isClickable: boolean) {
    // this.setState(() => ({
    //   clickable: isClickable,
    // }));
  }

  playSequence() {
    // const seq = this.state.sequence;
    // this.toggleClickable(false);
    // if (seq && seq.length) {
    //   seq.forEach((light: number): void => {
    //     setTimeout(
    //       this.setState(() => ({
    //         lightOn: light,
    //       })),
    //       500,
    //     );
    //   });
    // }
    // setTimeout(this.toggleClickable(true), this.props.counter * 1000);
  }

  render() {
    const lightOn = this.state.lightOn;
    return (
      <LightsWrapper>
        <LightsRow>
          <Light
            animate={lightOn === 0}
            topLeft={true}
            clickable={this.props.playerTurn}
            color="#00A74A"
            lightUpColor="#13ff7c"
            onClick={() => this.clickHandler(0)}
          />
          <Light
            animate={lightOn === 1}
            topRight={true}
            clickable={this.props.playerTurn}
            color="#9F0F17"
            lightUpColor="#ff4c4c"
            onClick={() => this.clickHandler(1)}
          />
        </LightsRow>
        <LightsRow>
          <Light
            animate={lightOn === 2}
            bottomLeft={true}
            clickable={this.props.playerTurn}
            color="#CCA707"
            lightUpColor="#fed93f"
            onClick={() => this.clickHandler(2)}
          />
          <Light
            animate={lightOn === 3}
            bottomRight={true}
            clickable={this.props.playerTurn}
            color="#094A8F"
            lightUpColor="#1c8cff"
            onClick={() => this.clickHandler(3)}
          />
        </LightsRow>
      </LightsWrapper>
    );
  }
}
