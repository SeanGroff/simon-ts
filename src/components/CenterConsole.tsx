import * as React from 'react';
import styled from 'styled-components';
import Counter from './Counter';
import StartStrictButton from './StartStrictButton';
import OnOffSwitch from './OnOffSwitch';

interface Props {
  hasMargin?: boolean;
  hasPadding?: boolean;
  power: boolean;
  counter: number;
  toggleGamePowerAction(power: boolean): { type: string; payload: boolean };
  startGameAction(): { type: string };
}

interface LabelProps {
  hasMargin?: boolean;
  hasPadding?: boolean;
}

const StyledWrapper = styled.div`
  position: absolute;
  border-radius: 100%;
  top: 50%;
  left: 50%;
  background-color: #ece7ee;
  border: 12px solid #333333;
  width: 220px;
  height: 220px;
  margin: -120px 0 0 -120px;
`;

const TopHalf = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: #333333;
  width: 220px;
  height: 110px;
  border-top-left-radius: 220px;
  border-top-right-radius: 220px;
  margin-top: -13px;
`;

const BottomHalf = TopHalf.extend`
  flex-direction: column;
  background-color: #ece7ee;
  border-top-left-radius: 0;
  border-top-right-radius: 0;
  border-bottom-left-radius: 220px;
  border-bottom-right-radius: 220px;
  margin-top: 0;
`;

const Header = styled.span`
  color: #ece7ee;
  font-family: 'Anton', sans-serif;
  font-size: 52px;
  letter-spacing: 2px;
  margin-top: 35px;
  text-transform: capitalize;
`;

const ButtonsRow = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-around;
`;

const ButtonWrapper = styled.span`
  display: flex;
  flex-direction: column;
  margin: 5px;
`;

const StyledLabel = styled.label`
  font-size: 14px;
  color: #222222;
  margin: ${({ hasMargin }: LabelProps) =>
    hasMargin ? 'auto 5px 0 5px' : '0'};
  padding: ${({ hasPadding }: LabelProps) => (hasPadding ? '0 0 5px 0' : '0')};
  text-transform: capitalize;
  text-align: center;
`;

export default (props: Props) => (
  <StyledWrapper>
    <TopHalf>
      <Header>{'simon'}</Header>
    </TopHalf>
    <BottomHalf>
      <ButtonsRow>
        <ButtonWrapper>
          <StyledLabel hasPadding={true}>{'count'}</StyledLabel>
          <Counter count={props.counter} on={props.power} />
        </ButtonWrapper>
        <ButtonWrapper>
          <StyledLabel hasPadding={true}>{'start'}</StyledLabel>
          <StartStrictButton
            startButton={true}
            disabled={!props.power || props.counter ? true : false}
            power={props.power}
            startGame={props.startGameAction}
          />
        </ButtonWrapper>
        <ButtonWrapper>
          <StyledLabel hasPadding={true}>{'strict'}</StyledLabel>
          <StartStrictButton
            startButton={false}
            disabled={!props.power}
            power={props.power}
          />
        </ButtonWrapper>
      </ButtonsRow>
      <ButtonsRow>
        <StyledLabel hasMargin={true}>{'off'}</StyledLabel>
        <OnOffSwitch
          power={props.power}
          powerOnOff={props.toggleGamePowerAction}
        />
        <StyledLabel hasMargin={true}>{'on'}</StyledLabel>
      </ButtonsRow>
    </BottomHalf>
  </StyledWrapper>
);
