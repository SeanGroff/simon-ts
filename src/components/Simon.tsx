import * as React from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';
import { toggleGamePowerAction } from '../actions/toggleGamePowerAction';
import { startGameThunk } from '../actions/startGameAction';
import { startNextRoundAction } from '../actions/startNextRoundAction';
// import { roundSuccessThunk } from '../actions/roundSuccessAction';
import LightsWrapper from './LightsWrapper';
import CenterConsole from './CenterConsole';
import { StoreState, Thunk } from '../types/types';

interface Props {
  power: boolean;
  counter: number;
  lightSequence: number[];
  playerTurn: boolean;
  toggleGamePowerAction(payload: boolean): { type: string; payload: boolean };
  startGameThunk(): Thunk;
  // roundSuccessThunk(): any;
  startNextRoundAction(): { type: string };
}

const SimonContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: #333333;
  border-radius: 100%;
  box-shadow: 0 0 12px #222222;
  width: 472px;
  height: 472px;
  position: relative;
`;

const mapStateToProps = (state: StoreState) => ({
  power: state.power,
  counter: state.counter,
  lightSequence: state.lightSequence,
  playerTurn: state.playerTurn,
});

const mapDispatchToProps = (dispatch: Dispatch<StoreState>) =>
  bindActionCreators(
    {
      toggleGamePowerAction,
      startGameThunk,
      // roundSuccessThunk,
      startNextRoundAction,
    },
    dispatch,
  );

export const Simon = (props: Props) => (
  <SimonContainer>
    <LightsWrapper
      power={props.power}
      counter={props.counter}
      lightSequence={props.lightSequence}
      playerTurn={props.playerTurn}
      toggleGamePowerAction={props.toggleGamePowerAction}
      startGameThunk={props.startGameThunk}
    />
    <CenterConsole
      power={props.power}
      counter={props.counter}
      toggleGamePowerAction={props.toggleGamePowerAction}
      startGameThunk={props.startGameThunk}
    />
  </SimonContainer>
);

const EnhancedSimon = connect(mapStateToProps, mapDispatchToProps)(Simon);

export default EnhancedSimon;
