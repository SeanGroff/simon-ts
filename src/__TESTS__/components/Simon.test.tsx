global.requestAnimationFrame = callback => {
  setTimeout(callback, 0);
};

import * as React from 'react';
import { shallow, configure } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import { Simon } from '../../components/Simon';

configure({ adapter: new Adapter() });

it('renders correctly', () => {
  const component = shallow(<Simon />);
  expect(component).toMatchSnapshot();
});
