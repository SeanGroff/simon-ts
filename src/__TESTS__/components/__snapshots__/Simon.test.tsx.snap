// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Lights
          counter={undefined}
          lightSequence={undefined}
          playerTurn={undefined}
          power={undefined}
          startGameThunk={undefined}
          toggleGamePowerAction={undefined}
/>,
        <Unknown
          counter={undefined}
          power={undefined}
          startGameThunk={undefined}
          toggleGamePowerAction={undefined}
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "class",
        "props": Object {
          "counter": undefined,
          "lightSequence": undefined,
          "playerTurn": undefined,
          "power": undefined,
          "startGameThunk": undefined,
          "toggleGamePowerAction": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "function",
        "props": Object {
          "counter": undefined,
          "power": undefined,
          "startGameThunk": undefined,
          "toggleGamePowerAction": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Lights
            counter={undefined}
            lightSequence={undefined}
            playerTurn={undefined}
            power={undefined}
            startGameThunk={undefined}
            toggleGamePowerAction={undefined}
/>,
          <Unknown
            counter={undefined}
            power={undefined}
            startGameThunk={undefined}
            toggleGamePowerAction={undefined}
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "class",
          "props": Object {
            "counter": undefined,
            "lightSequence": undefined,
            "playerTurn": undefined,
            "power": undefined,
            "startGameThunk": undefined,
            "toggleGamePowerAction": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "function",
          "props": Object {
            "counter": undefined,
            "power": undefined,
            "startGameThunk": undefined,
            "toggleGamePowerAction": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
